worker/src/websocket.rs

```diff
@@ -0,0 +1,267 @@
+use crate::ws_events::{CloseEvent, ErrorEvent, MessageEvent};
+use crate::{Error, Result};
+use serde::Serialize;
+use std::future::Future;
+use wasm_bindgen::convert::FromWasmAbi;
+use wasm_bindgen::prelude::Closure;
+use wasm_bindgen::{JsCast, JsValue};
+
+/// Struct holding the values for a JavaScript `WebSocketPair`
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct WebSocketPair {
+    pub client: WebSocket,
+    pub server: WebSocket,
+}
+
+impl WebSocketPair {
+    /// Creates a new `WebSocketPair`.
+    pub fn new() -> Result<Self> {
+        let mut pair = worker_sys::WebSocketPair::new();
+        let client = pair.client()?.into();
+        let server = pair.server()?.into();
+        Ok(Self { client, server })
+    }
+}
+
+/// Wrapper struct for underlying worker-sys `WebSocket`
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct WebSocket {
+    socket: worker_sys::WebSocket,
+}
+
+impl WebSocket {
+    /// Accept the server-side of a websocket connection
+    pub fn accept(&self) -> Result<()> {
+        self.socket.accept().map_err(Error::from)
+    }
+
+    /// Serialize data into a string using serde and send it through the `WebSocket`
+    pub fn send<T: Serialize>(&self, data: &T) -> Result<()> {
+        let value = serde_json::to_string(data)?;
+        self.send_with_str(value.as_str())
+    }
+
+    /// Sends a raw string through the `WebSocket`
+    pub fn send_with_str<S: AsRef<str>>(&self, data: S) -> Result<()> {
+        self.socket
+            .send_with_str(data.as_ref())
+            .map_err(Error::from)
+    }
+
+    /// Sends raw binary data through the `WebSocket`.
+    pub fn send_with_binary(&self, bytes: Vec<u8>) -> Result<()> {
```

Comment:
We probably shouldn't require a `Vec<u8>` here, someone might only have a reference to a byte slice. I suggest changing `bytes` to either `&[u8]` or `impl AsRef<[u8]>`.

Reply:
Wasn't that an issue with the response referencing to released memory addresses?
This was a slice before, but to avoid the same error, I've used an owned array there!

Reply:
That was unrelated, because we gave javascript iself the reference, I'm talking about taking using a slice in rust and then copying that into the `Uint8Array`. So the code would look like
```rust
/// Sends raw binary data through the `WebSocket`.
pub fn send_with_binary(&self, bytes: impl AsRef<[u8]>) -> Result<()> {
    let bytes = bytes.as_ref();
    let array = js_sys::Uint8Array::new_with_length(bytes.len() as u32);
    array.copy_from(bytes);
    self.socket.send_with_u8_array(array).map_err(Error::from)
}
```